sale_rental/rental.py

```diff
@@ -0,0 +1,636 @@
+
+    @api.model
+    def _get_rental_date_planned(self, line):
+        return line.start_date
+
+    @api.model
+    def _prepare_order_line_procurement(
+            self, order, line, group_id=False):
+        res = super(SaleOrder, self)._prepare_order_line_procurement(
+            order, line, group_id=group_id)
+        if (
+                line.product_id.rented_product_id and
+                line.rental_type == 'new_rental'):
+            res.update({
+                'product_id': line.product_id.rented_product_id.id,
+                'product_qty': line.rental_qty,
+                'product_uos_qty': line.rental_qty,
+                'product_uom': line.product_id.rented_product_id.uom_id.id,
+                'product_uos': line.product_id.rented_product_id.uom_id.id,
+                'location_id':
+                order.warehouse_id.rental_out_location_id.id,
+                'route_ids':
+                [(6, 0, [line.order_id.warehouse_id.rental_route_id.id])],
+                'date_planned': self._get_rental_date_planned(line),
+                })
+        elif line.sell_rental_id:
+            res['route_ids'] = [(6, 0, [
+                line.order_id.warehouse_id.sell_rented_product_route_id.id])]
+        return res
+
+    @api.model
+    def _prepare_rental(self, so_line):
+        return {'start_order_line_id': so_line.id}
+
+    @api.multi
+    def action_button_confirm(self):
+        res = super(SaleOrder, self).action_button_confirm()
+        for order in self:
+            for line in order.order_line:
+                if line.rental_type == 'new_rental':
+                    self.env['sale.rental'].create(self._prepare_rental(line))
+                elif line.rental_type == 'rental_extension':
+                    line.extension_rental_id.in_move_id.date_expected =\
+                        line.end_date
+                    line.extension_rental_id.in_move_id.date = line.end_date
+                elif line.sell_rental_id:
+                    if line.sell_rental_id.out_move_id.state != 'done':
+                        raise Warning(
+                            _('Cannot sell the rental %s because it has '
+                                'not been delivered')
+                            % line.sell_rental_id.display_name)
+                    line.sell_rental_id.in_move_id.action_cancel()
+        return res
+
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    rental = fields.Boolean(string='Rental')
+    can_sell_rental = fields.Boolean(string='Can Sell from Rental')
+    rental_type = fields.Selection([
+        ('new_rental', 'New Rental'),
+        ('rental_extension', 'Rental Extension'),
+        ], 'Rental Type',
+        readonly=True, states={'draft': [('readonly', False)]})
+    extension_rental_id = fields.Many2one(
+        'sale.rental', string='Rental to Extend')
+    rental_qty = fields.Float(
+        string='Rental Quantity', digits=dp.get_precision('Product UoS'))
```

Comment:
I find confusing to see the following fields in the sale order line:
Number of Days
Rental Quantity
Quantity

I suggest that adding a tooltip to the field 'Rental Quantity' will help the user to clarify what to enter here.
For example:
help = "Indicate the number of items that will be delivered for rental."